lib/arithmetic.js

```diff
@@ -159,12 +228,22 @@ exports.traceExpression = function(builder, expression, traceOpts) {
   // expand any globbing operations in the expression
   var expandPromise = Promise.resolve();
   operations.forEach(function(operation) {
-    expandPromise = expandPromise.then(function() {
-      return Promise.resolve(expandGlobAndCanonicalize(builder, operation))
-      .then(function(expanded) {
-        expandedOperations = expandedOperations.concat(expanded);
+    if (operation.bulkOperation) {
+      var expandedTreePromise = traceExpression(builder, operation.bulkOperation, traceOpts)
+      expandPromise = expandPromise.then(function() {
+        return Promise.resolve(expandedTreePromise)
+               .then(function(fullTree){
+                 expandedOperations = expandedOperations.concat({ preExpandedTree: fullTree, operator: operation.operator })
```

Comment:
To clarify the reason being here that each "step" of the algorithm should have a clear purpose. To do globbing and paren expansion together is an arbitrary conflation just to take advantage of the same loop, something which can rather be split up into two loops (corresponding to either (1) or (2)).